File: Python/LeetCode/287.py
class Solution(object):
    def findDuplicate1(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        for index in range(len(nums)):
            while nums[index] != index:
                if nums[index] == nums[nums[index]]:
                    return nums[index]
                else:
                    temp = nums[index]
                    nums[index] = nums[temp]
                    nums[temp] = temp

    def findDuplicate2(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        slow = nums[0]
        fast = nums[nums[0]]

        while nums[slow] != nums[fast]:
            slow = nums[slow]
            fast = nums[nums[fast]]

        head = 0
        while nums[slow] != nums[head]:
            slow = nums[slow]
            head = nums[head]

        return nums[head]

    def findDuplicate3(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums = sorted(nums)
        index = 0

        while index < len(nums) - 1:
            if nums[index] == nums[index + 1]:
                return nums[index]
            index += 1

File: Cpp/LeetCode/287.cpp
//
// Created by caizixiang on 2019-07-28.
//

class Solution {
public:
    int findDuplicate1(vector<int>& nums) {
        int slow = nums[0];
        int fast = nums[nums[0]];

        while (nums[slow] != nums[fast])
        {
            slow = nums[slow];
            fast = nums[nums[fast]];
        }

        int head = 0;
        while (nums[head] != nums[slow])
        {
            head = nums[head];
            slow = nums[slow];
        }

        return nums[head];
    }

    int findDuplicate2(vector<int>& nums) {
        int tail = nums.size() - 1, head = 0, mid;
        while (head <= tail)
        {
            mid = (head + tail) >> 1;
            int count = 0;
            for (int num: nums)
            {
                if (num <= mid)
                    count++;
            }

            if (count>mid)
                tail = mid - 1;
            else
                head = mid + 1;
        }

        return head;
    }
};